jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java

```diff
@@ -145,307 +100,108 @@ public ContextProvider()
-            // order all jetty.deploy.environmentXml[.zzzz] properties lexically
-            // apply the context xml files named by the ordered jetty.deploy.environmentXml[.zzzz] properties
-            String env = app.getEnvironmentName() == null ? "" : app.getEnvironmentName();
 
-            if (StringUtil.isNotBlank(env))
+            // check if there is a specific ContextHandler type to create set in the
+            // properties associated with the webapp. If there is, we create it _before_
+            // applying the environment xml file.
+            String contextHandlerClassName = (String)appAttributes.getAttribute(Deployable.CONTEXT_HANDLER_CLASS);
+            if (contextHandlerClassName != null)
             {
-                List<Path> envPropertyFiles = new ArrayList<>();
-                Path parent = app.getPath().getParent();
-
-                //Get all environment specific properties files for this environment,
-                //order them according to the lexical ordering of the filenames
-                try (Stream<Path> paths = Files.list(parent))
+                Class<?> contextClass = Loader.loadClass(contextHandlerClassName);
+                if (contextClass == null)
                 {
-                    envPropertyFiles = paths.filter(Files::isRegularFile)
-                        .map(p -> parent.relativize(p))
-                        .filter(p ->
-                        {
-                            String name = p.getName(0).toString();
-                            if (!name.endsWith(".properties"))
-                                return false;
-                            if (!name.startsWith(env))
-                                return false;
-                            return true;
-                        }).sorted().collect(Collectors.toList());
+                    throw new IllegalStateException("Unknown ContextHandler class " + contextHandlerClassName + " for " + app + " in environment " + environment.getName());
                 }
-
-                if (LOG.isDebugEnabled())
-                    LOG.debug("Environment property files {}", envPropertyFiles);
-
-                Map<String, Path> envXmlFilenameMap = new HashMap<>();
-                for (Path file : envPropertyFiles)
+                else
                 {
-                    Path resolvedFile = parent.resolve(file);
-                    if (Files.exists(resolvedFile))
-                    {
-                        Properties tmp = new Properties();
-                        try (InputStream stream = Files.newInputStream(resolvedFile))
-                        {
-                            tmp.load(stream);
-                            //put each property into our substitution pool
-                            tmp.stringPropertyNames().forEach(k -> properties.put(k, tmp.getProperty(k)));
-                        }
-                    }
+                    context = contextClass.getDeclaredConstructor().newInstance();
```

Comment:
Constructor should be `public`, so use:

```suggestion
                    context = contextClass.getConstructor().newInstance();
```

I'd also simplify this `if/else` by removing the `else`:

```java
if (contextClass == null)
    throw ...;
context = contextClass.getConstructor().newInstance();